group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -0,0 +1,448 @@
+                request.groupId(),
+                request.generationIdOrMemberEpoch() < 0
+            );
+        } catch (GroupIdNotFoundException ex) {
+            // Maintain backward compatibility. This is a bit weird in the
+            // context of the new protocol though.
+            throw Errors.ILLEGAL_GENERATION.exception();
+        }
+
+        // Validate the request based on the group type.
+        switch (group.type()) {
+            case GENERIC:
+                validateOffsetCommitForGenericGroup(
+                    (GenericGroup) group,
+                    request
+                );
+                break;
+
+            case CONSUMER:
+                validateOffsetCommitForConsumerGroup(
+                    (ConsumerGroup) group,
+                    context,
+                    request
+                );
+                break;
+        }
+    }
+
+    /**
+     * Validates an OffsetCommit request for a generic group.
+     *
+     * @param group     The generic group.
+     * @param request   The actual request.
+     */
+    public void validateOffsetCommitForGenericGroup(
+        GenericGroup group,
+        OffsetCommitRequestData request
+    ) throws KafkaException {
+        if (group.isInState(DEAD)) {
+            throw Errors.COORDINATOR_NOT_AVAILABLE.exception();
+        }
+
+        if (request.generationIdOrMemberEpoch() < 0 && group.isInState(EMPTY)) {
+            // When the generation id is -1, the request comes from either the admin client
+            // or a consumer which does not use the group management facility. In this case,
+            // the request can commit offsets if the group is empty.
+            return;
+        }
+
+        Optional<String> groupInstanceId = OffsetCommitRequest.groupInstanceId(request);
+        if (request.generationIdOrMemberEpoch() >= 0 || !request.memberId().isEmpty() || groupInstanceId.isPresent()) {
+            // We are validating three things:
+            // 1. If the `groupInstanceId` is present, then it exists and is mapped to `memberId`;
+            // 2. The `memberId` exists in the group; and
+            // 3. The `generationId` matches the current generation id.
+            if (groupInstanceId.isPresent()) {
+                String memberId = group.staticMemberId(groupInstanceId.get());
+                if (memberId == null) {
+                    throw Errors.UNKNOWN_MEMBER_ID.exception();
+                } else if (!request.memberId().equals(memberId)) {
+                    throw Errors.FENCED_INSTANCE_ID.exception();
+                }
+            }
+
+            if (!group.hasMemberId(request.memberId())) {
+                throw Errors.UNKNOWN_MEMBER_ID.exception();
+            }
+
+            if (request.generationIdOrMemberEpoch() != group.generationId()) {
+                throw Errors.ILLEGAL_GENERATION.exception();
+            }
+        } else if (!group.isInState(EMPTY)) {
```

Comment:
>     // If the request does not contain the member id and the generation
            // id (version 0), offset commits are only accepted when the group
            // is not empty.

Does this mean we should throw the error if the group is empty?

Reply:
The comment seems wrong. It should be `is empty`.